### PR TITLE
refactor(composer): Simplify Packagist lookups

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

- Introduce schemas and transformations for Composer datasource
- Ensure `homepage` and `sourceUrl` are extracted from the latest release

## Context

- Ref: #19495

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — APPROVED** [review]

**Comment by bobvandevijver** [comment]
@rarkins Can this be reverted? Almost all our composer dependencies now fail to be resolved correctly, causing all our open MRs to be closed with the following logging: 

```
{"name":"renovate","hostname":"runner-3cixccwz-project-43-concurrent-0","pid":15,"level":20,"logContext":"xsuz73R5TD3hKe_xpfZnl","repository":"living-textbook/living-textbook","datasource":"packagist","packageName":"doctrine/collections","err":{"message":"Invalid Version: 3.0.x-dev","stack":"TypeError: Invalid Version: 3.0.x-dev\n    at new SemVer (/opt/buildpack/tools/renovate/34.100.2/node_modules/semver/classes/semver.js:38:13)\n    at compare (/opt/buildpack/tools/renovate/34.100.2/node_modules/semver/functions/compare.js:3:3)\n    at Object.gt [as isGreaterThan] (/opt/buildpack/tools/renovate/34.100.2/node_modules/semver/functions/gt.js:2:29)\n    at Object.isGreaterThan (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/modules/versioning/composer/index.ts:114:14)\n    at Object.transform (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/modules/datasource/packagist/schema.ts:47:41)\n    at ZodEffects._parse (/opt/buildpack/tools/renovate/34.100.2/node_modules/zod/lib/types.js:2613:39)\n    at ZodEffects._parseSync (/opt/buildpack/tools/renovate/34.100.2/node_modules/zod/lib/types.js:111:29)\n    at ZodEffects.safeParse (/opt/buildpack/tools/renovate/34.100.2/node_modules/zod/lib/types.js:141:29)\n    at ZodEffects.parse (/opt/buildpack/tools/renovate/34.100.2/node_modules/zod/lib/types.js:122:29)\n    at PackagistDatasource.packagistOrgLookup (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/modules/datasource/packagist/index.ts:214:43)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at /opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/util/cache/package/decorator.ts:78:20\n    at PackagistDatasource.getReleases (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/modules/datasource/packagist/index.ts:230:33)\n    at getRegistryReleases (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/modules/datasource/index.ts:73:15)\n    at huntRegistries (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/modules/datasource/index.ts:110:13)\n    at fetchReleases (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/modules/datasource/index.ts:296:15)\n    at getPkgReleases (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/modules/datasource/index.ts:353:7)\n    at lookupUpdates (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/workers/repository/process/lookup/index.ts:82:26)\n    at fetchDepUpdates (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/workers/repository/process/fetch.ts:59:15)\n    at /opt/buildpack/tools/renovate/34.100.2/node_modules/p-map/index.js:57:22"},"msg":"Datasource unknown error","time":"2023-01-13T10:20:22.934Z","v":0}
{"name":"renovate","hostname":"runner-3cixccwz-project-43-concurrent-0","pid":15,"level":20,"logContext":"xsuz73R5TD3hKe_xpfZnl","repository":"living-textbook/living-textbook","dependency":"doctrine/collections","packageFile":"composer.json","msg":"Failed to look up packagist dependency doctrine/collections","time":"2023-01-13T10:20:22.934Z","v":0}
```

**Comment by rarkins** [comment]
In progress: https://github.com/renovatebot/renovate/pull/19824

@zharinov is the above log enough to work off, or do you need more logs or reproduction?

File: lib/modules/datasource/packagist/schema.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        for (const composerV2Release of responsePackage) {
+          const { version, time: releaseTimestamp } = composerV2Release;
+          const dep: Release = {
+            version: version.replace(/^v/, ''),
+            gitRef: version,
+            releaseTimestamp,
+          };
+          releases.push(dep);
+
+          if (!versioning.isValid(version)) {
```
Comment (viceice) [inline lib/modules/datasource/packagist/schema.ts:43]: you probably need to check `versioning.isSingleVersion(version)` 🤔

**Comment by herndlm** [comment]
In case you need more logs for debugging (please ignore and forgive me in case you don't :))

```
  DEBUG: Datasource unknown error (repository=<REDACTED>)
         "datasource": "packagist",
         "packageName": "mustangostang/spyc",
         "err": {
           "issues": [
             {
               "validation": "url",
               "code": "invalid_string",
               "message": "Invalid url",
               "path": [0, "packages", "mustangostang/spyc", 0, "source", "url"]
             }
           ],
           "addIssue": "[function]",
           "addIssues": "[function]",
           "name": "ZodError",
           "message": "[\n  {\n    \"validation\": \"url\",\n    \"code\": \"invalid_string\",\n    \"message\": \"Invalid url\",\n    \"path\": [\n      0,\n      \"packages\",\n      \"mustangostang/spyc\",\n      0,\n      \"source\",\n      \"url\"\n    ]\n  }\n]",
           "stack": "ZodError: [\n  {\n    \"validation\": \"url\",\n    \"code\": \"invalid_string\",\n    \"message\": \"Invalid url\",\n    \"path\": [\n      0,\n      \"packages\",\n      \"mustangostang/spyc\",\n      0,\n      \"source\",\n      \"url\"\n    ]\n  }\n]\n    at handleResult (/opt/buildpack/tools/renovate/34.100.2/node_modules/zod/lib/types.js:29:23)\n    at ZodEffects.safeParse (/opt/buildpack/tools/renovate/34.100.2/node_modules/zod/lib/types.js:142:16)\n    at ZodEffects.parse (/opt/buildpack/tools/renovate/34.100.2/node_modules/zod/lib/types.js:122:29)\n    at PackagistDatasource.packagistOrgLookup (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/modules/datasource/packagist/index.ts:214:43)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at /opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/util/cache/package/decorator.ts:78:20\n    at PackagistDatasource.getReleases (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/modules/datasource/packagist/index.ts:230:33)\n    at getRegistryReleases (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/modules/datasource/index.ts:73:15)\n    at huntRegistries (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/modules/datasource/index.ts:110:13)\n    at fetchReleases (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/modules/datasource/index.ts:296:15)\n    at getPkgReleases (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/modules/datasource/index.ts:353:7)\n    at lookupUpdates (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/workers/repository/process/lookup/index.ts:82:26)\n    at fetchDepUpdates (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/workers/repository/process/fetch.ts:59:15)\n    at /opt/buildpack/tools/renovate/34.100.2/node_modules/p-map/index.js:57:22"
         }
  DEBUG: Failed to look up packagist dependency mustangostang/spyc (repository=<REDACTED>, packageFile=<REDACTED>/composer.json, dependency=mustangostang/spyc)
```

```
DEBUG: Datasource unknown error (repository=<REDACTED>)
         "datasource": "packagist",
         "packageName": "laminas/laminas-db",
         "err": {
           "issues": [
             {
               "code": "invalid_type",
               "expected": "string",
               "received": "undefined",
               "path": [0, "packages", "laminas/laminas-db", 11, "time"],
               "message": "Required"
             }
           ],
           "addIssue": "[function]",
           "addIssues": "[function]",
           "name": "ZodError",
           "message": "[\n  {\n    \"code\": \"invalid_type\",\n    \"expected\": \"string\",\n    \"received\": \"undefined\",\n    \"path\": [\n      0,\n      \"packages\",\n      \"laminas/laminas-db\",\n      11,\n      \"time\"\n    ],\n    \"message\": \"Required\"\n  }\n]",
           "stack": "ZodError: [\n  {\n    \"code\": \"invalid_type\",\n    \"expected\": \"string\",\n    \"received\": \"undefined\",\n    \"path\": [\n      0,\n      \"packages\",\n      \"laminas/laminas-db\",\n      11,\n      \"time\"\n    ],\n    \"message\": \"Required\"\n  }\n]\n    at handleResult (/opt/buildpack/tools/renovate/34.100.2/node_modules/zod/lib/types.js:29:23)\n    at ZodEffects.safeParse (/opt/buildpack/tools/renovate/34.100.2/node_modules/zod/lib/types.js:142:16)\n    at ZodEffects.parse (/opt/buildpack/tools/renovate/34.100.2/node_modules/zod/lib/types.js:122:29)\n    at PackagistDatasource.packagistOrgLookup (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/modules/datasource/packagist/index.ts:214:43)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at /opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/util/cache/package/decorator.ts:78:20\n    at PackagistDatasource.getReleases (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/modules/datasource/packagist/index.ts:230:33)\n    at getRegistryReleases (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/modules/datasource/index.ts:73:15)\n    at huntRegistries (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/modules/datasource/index.ts:110:13)\n    at fetchReleases (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/modules/datasource/index.ts:296:15)\n    at getPkgReleases (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/modules/datasource/index.ts:353:7)\n    at lookupUpdates (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/workers/repository/process/lookup/index.ts:82:26)\n    at fetchDepUpdates (/opt/buildpack/tools/renovate/34.100.2/node_modules/renovate/lib/workers/repository/process/fetch.ts:59:15)\n    at /opt/buildpack/tools/renovate/34.100.2/node_modules/p-map/index.js:57:22"
         }
  DEBUG: Failed to look up packagist dependency laminas/laminas-db (repository=<REDACTED>, packageFile=<REDACTED>/composer.json, dependency=laminas/laminas-db)
```

**Pull request merged**